### PR TITLE
MANTA-4956 Rebalancer: Investigate the use of workspaces in cargo to separate out the manager and agent

### PR DESCRIPTION
**Pull request by rhb2**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -49,15 +49,13 @@ include ./deps/eng/tools/mk/Makefile.smf.defs
...
+	$(CARGO) build --manifest-path=agent/Cargo.toml --release
+	$(CARGO) build --manifest-path=manager/Cargo.toml --release
 
 debug:
-	$(CARGO) build --bin rebalancer-agent
-	$(CARGO) build --bin rebalancer-manager --features "postgres"
-	$(CARGO) build --bin rebalancer-adm --features "postgres"
-	cp src/config.json target/debug/
+	$(CARGO) build --manifest-path=agent/Cargo.toml
+	$(CARGO) build --manifest-path=manager/Cargo.toml
```
Comment (rjloura) [inline Makefile]: Can these just be:
```
cargo build --release
```
and
```
cargo build
```

File: agent/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+base64 = "0.10.1"
+futures = "0.1.27"
+gotham = "0.3.0"
+gotham_derive = "0.3.0"
+hyper = "0.12"
+joyent-rust-utils = { git = "https://github.com/joyent/rust-utils" }
+lazy_static = "1.4.0"
+libmanta = { git = "https://github.com/joyent/rust-libmanta", tag = "v0.7.0" }
+mime = "0.3.13"
+moray = { git = "https://github.com/joyent/rust-moray", tag = "v0.9.0" }
```
Comment (rjloura) [inline agent/Cargo.toml]: Can you update this to 0.9.2

File: agent/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+# No features by default
+default = []
+
+[dependencies]
+base64 = "0.10.1"
+futures = "0.1.27"
+gotham = "0.3.0"
+gotham_derive = "0.3.0"
+hyper = "0.12"
+joyent-rust-utils = { git = "https://github.com/joyent/rust-utils" }
```
Comment (rjloura) [inline agent/Cargo.toml]: Can you add `tag = "v0.2.0"` to this?

File: manager/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+clap = "2.33.0"
+crossbeam-channel = "0.3.8"
+crossbeam-deque = "0.7.1"
+futures = "0.1.27"
+hyper = "0.12"
+Inflector = "0.11.4"
+lazy_static = "1.4.0"
+libmanta = { git = "https://github.com/joyent/rust-libmanta", tag = "v0.7.0" }
+mime = "0.3.13"
+moray = { git = "https://github.com/joyent/rust-moray", tag = "v0.9.0" }
```
Comment (rjloura) [inline manager/Cargo.toml]: v0.9.2

File: manager/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+sharkspotter = { git = "https://github.com/joyent/rust-sharkspotter", tag = "v0.10.0" }
```
Comment (rjloura) [inline manager/Cargo.toml]: v0.10.2

File: manager/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,52 @@
+[package]
+name = "manager"
+version = "0.1.0"
+authors = ["Jenkins Agent <jenkins-agent@joyent.com>"]
+edition = "2018"
+workspace = ".."
+
+[features]
```
Comment (rjloura) [inline manager/Cargo.toml]: I wonder if we need this anymore.  I think we may be able to put the features inline with their dependencies.

File: manager/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+crossbeam-deque = "0.7.1"
+futures = "0.1.27"
+hyper = "0.12"
+Inflector = "0.11.4"
+lazy_static = "1.4.0"
+libmanta = { git = "https://github.com/joyent/rust-libmanta", tag = "v0.7.0" }
+mime = "0.3.13"
+moray = { git = "https://github.com/joyent/rust-moray", tag = "v0.9.0" }
+sharkspotter = { git = "https://github.com/joyent/rust-sharkspotter", tag = "v0.10.0" }
+diesel = {version = "1.4.2", features = ["sqlite"]}
```
Comment (rjloura) [inline manager/Cargo.toml]: I don't think we use sqlite in the manager anymore.

File: agent/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+serde = { version = "1.0.91", features = ["derive"] }
+serde_derive = "1.0.91"
+serde_json = "1.0.39"
+threadpool = "1.7.1"
+url = "1.7"
+uuid = { version = "0.7.4", features = ["v4"] }
+walkdir = "2"
+
+[[bin]]
+name = "rebalancer-agent"
```
Comment (rjloura) [inline agent/Cargo.toml]: Do we need this anymore?  Or is it just used to get the name we want for the agent "rebalancer-agent"

File: manager/src/main.rs [hunk 9]
```diff
@@ -338,7 +350,8 @@ fn main() {
 
     let config = config::Config::parse_config(config_file)
         .map_err(|e| {
-            remora::error!("Error parsing config: {}", e);
+            //xxxremora::error!("Error parsing config: {}", e);
```
Comment (rjloura) [inline manager/src/main.rs]: remove comment.

File: rebalancer/src/common.rs [hunk 10]
```diff
@@ -264,7 +263,8 @@ impl FromSql<sql_types::Text, Pg> for ObjectSkippedReason {
     }
 }
 
-pub(crate) fn get_sharks_from_value(
+//pub(crate) fn get_sharks_from_value(
```
Comment (rjloura) [inline rebalancer/src/common.rs]: remove comment

File: rebalancer/src/common.rs [hunk 11]
```diff
@@ -281,7 +281,8 @@ pub(crate) fn get_sharks_from_value(
 }
 
 #[allow(non_snake_case)]
-pub(crate) fn get_objectId_from_value(
+//pub(crate) fn get_objectId_from_value(
```
Comment (rjloura) [inline rebalancer/src/common.rs]: remove comment

File: manager/Cargo.toml (hunk 6 above)
Comment (rhb2) [reply manager/Cargo.toml]: Looks like we don't need it now.

File: manager/Cargo.toml (hunk 7 above)
Comment (rhb2) [reply manager/Cargo.toml]: Done.

File: agent/Cargo.toml (hunk 8 above)
Comment (rhb2) [reply agent/Cargo.toml]: Yeah this basically just tells Cargo what to name the binary.  If we don't supply this section, cargo will name it after the crate itself which is just "agent" which I think we don't want.

File: manager/src/main.rs (hunk 9 above)
Comment (rhb2) [reply manager/src/main.rs]: Done.

File: rebalancer/src/common.rs (hunk 10 above)
Comment (rhb2) [reply rebalancer/src/common.rs]: Done.

File: rebalancer/src/common.rs (hunk 11 above)
Comment (rhb2) [reply rebalancer/src/common.rs]: Done.

File: manager/Cargo.toml (hunk 5 above)
Comment (rhb2) [reply manager/Cargo.toml]: Done.

File: manager/Cargo.toml (hunk 4 above)
Comment (rhb2) [reply manager/Cargo.toml]: Done.

File: agent/Cargo.toml (hunk 3 above)
Comment (rhb2) [reply agent/Cargo.toml]: Done.  Obviously, this introduces some minor changes to Cargo.lock in the root as well.

File: agent/Cargo.toml (hunk 2 above)
Comment (rhb2) [reply agent/Cargo.toml]: Done.

File: Makefile (hunk 1 above)
Comment (rhb2) [reply Makefile]: Looks like we need to build them as separate crates in order to ensure that the agent does not depend on libpq.so

Comment (rjloura) [reply Makefile]: bummer.  Good catch though.

File: rebalancer/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+base64 = "0.10.1"
+clap = "2.33.0"
+crossbeam-channel = "0.3.8"
+diesel = {version = "1.4.2", features = ["sqlite"]}
+futures = "0.1.27"
+uuid = { version = "0.7.4", features = ["v4"] }
+gotham = "0.3.0"
+gotham_derive = "0.3.0"
+hyper = "0.12"
+joyent-rust-utils = { git = "https://github.com/joyent/rust-utils" }
```
Comment (rjloura) [inline rebalancer/Cargo.toml]: This one too

Comment (rhb2) [reply rebalancer/Cargo.toml]: Done.

**Pull request merged**